### PR TITLE
Avatar menu + Sign Out

### PR DESCRIPTION
**Pull request by jwicks31**
Part of #93 

### Product Spec
screen shot 2018-08-01 at 6 56 39 pm

### Description of Changes

* [x] Installs @material-ui/core
* [x] Sign Out Action Creator
* [x] Sign Out Added to User Reducer
* [x] SignOutSage redirect to landing page
* [x] Avatar Menu Storybook Component
* [x] AvatarMenu Component
* [x] Saga/Reducer Unit Tests
* [x] Update Navbar to render AvatarMenu when User is logged in and verified

The styling for the avatar menu + the profile/settings pages still need to be set up. This is hidden behind a feature flag and does add the Sign Out functionality. Would like to merge in to keep PR's smaller.

### Feature Toggle

`?ft=login,auth,avatar` - To login and show avatar after complete
`?ft=avatar,auth` - Shows avatar drop down menu when signed in. Logout is the only link connected.

### PR Review Checklist

* [ ] **Do the changes match the product spec?**
* [ ] **Are there changes in this PR that weren't originally in the product spec?**
* [ ] **Is a feature toggle introduced for this PR? Is it disabled by default?**
* [ ] **Have you tested the changes?**
* [ ] **Is QE review required? If so, has this been reviewed by QE?** (For all new features, review by QE is required. Minor changes which already have passing functional tests may not need a new QE review -- when in doubt, loop in a QE reviewer).
* [ ] **Are there code changes that require new unit tests? If so, are those tests included?** (A separate PR for tests is not appropriate).
* [ ] **Are documentation updates required? If so, does this PR include it?** (Separate documentation PR is generally not appropriate).
* [ ] **Are the added/edited files named appropriately and located correctly in the project directory structure?**

### Warning

Merging to `master` automatically triggers a deploy in production.

### CODE REVIEW

File: src/components/atoms/AvatarMenu/AvatarMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  anchorEl: object
+}
+
+export class AvatarMenu extends React.Component<AvatarMenuProps, AvatarMenuState> {
+  state: { anchorEl: any } = {
+    anchorEl: null,
+  }
+
+  handleClick = (event: any) => {
+    this.setState({ anchorEl: event.currentTarget })
```
Comment (warrenv) [inline src/components/atoms/AvatarMenu/AvatarMenu.tsx]: We should use the functional form of setState so that our state updates are always deterministic.
See https://medium.com/javascript-scene/setstate-gate-abc10a9b2d82

Comment (jwicks31) [reply src/components/atoms/AvatarMenu/AvatarMenu.tsx]: Is this still true even though we aren't concerned about the previous state?

Comment (warrenv) [reply src/components/atoms/AvatarMenu/AvatarMenu.tsx]: I'm thinking more in terms of the whole codebase rather than this particular usage.
We want to adopt a consistent way of using setState throughout the codebase. By using the functional form, we avoid any instance, now or in the future, where non-deterministic behavior is the root of a bug. The time and expense are completely avoided by choosing the functional form.

Comment (wzalazar) [reply src/components/atoms/AvatarMenu/AvatarMenu.tsx]: I think there is another possibility to handle states with the render props.
I talked about it with @lautarodragan 2 months ago

I leave  this article 

https://reactjs.org/docs/render-props.html

**Comment by jwicks31** [comment]
I separated out the Navbar component into a container and stateless component in #169. I wanted to better be able to unit test the Navbar component, but didn't want to commit it directly into this PR without Review.

File: src/components/atoms/AvatarMenu/AvatarMenu.tsx (hunk 1 above)
Comment (jwicks31) [reply src/components/atoms/AvatarMenu/AvatarMenu.tsx]: I updated it to just the functional setState for this PR. I've created another branch and added an AnchorProvider Render Prop Component. Once we merge these 2 PR's in I will create that PR.

**Comment by jwicks31** [comment]
@warrenv @wzalazar @lautarodragan could you please review this PR whenever you get a chance?

File: src/reducers/User.test.ts [hunk 2]
```diff
@@ -43,13 +46,29 @@ describe('user reducer', async (should: any) => {
...
         verified: true,
         createdAt: 'test',
+        avatar: Images.Avatar,
       },
     }),
   })
 
   assert({
     given: 'default state and PROFILE_SUCCESS action with user',
     should: 'state with payload as user',
```
Comment (warrenv) [inline src/reducers/User.test.ts]: 'return state ...' reads better here don't you agree?

File: src/reducers/User.ts [hunk 3]
```diff
@@ -35,6 +37,8 @@ export const user = (state: any, action: any) => {
           token: action.payload.token,
         },
       }
+    case Actions.SignOut.SIGN_OUT:
+      return {}
```
Comment (warrenv) [inline src/reducers/User.ts]: Maybe returning defaultState would be better?

Comment (jwicks31) [reply src/reducers/User.ts]: Fixed this

**Review by warrenv — DISMISSED** [review]

File: src/components/molecules/Navbar/Navbar.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  readonly displaySearch?: boolean
+  readonly searchShadow?: boolean
+  readonly user?: User
+}
+
+const getSearchQuery = () => {
+  const currentLocation = browserHistory.getCurrentLocation()
+
+  if (currentLocation.pathname === '/works') return (currentLocation.query as any).query
+  else return ''
```
Comment (wzalazar) [inline src/components/molecules/Navbar/Navbar.tsx]: I think is not necessary the else sentence

**Comment by wzalazar** [comment]
@jwicks31 is ready to review?

**Comment by jwicks31** [comment]
@wzalazar it is!

**Comment by wzalazar** [comment]
image

I have doubts about what we should expect regarding the UI... Because it's different

**Comment by jwicks31** [comment]
@wzalazar I left this note in the initial comment section: "The styling for the avatar menu + the profile/settings pages still need to be set up. This is hidden behind a feature flag and does add the Sign Out functionality. Would like to merge in to keep PR's smaller." 

However, if we don't want to do that I can complete the style updates in this PR.

File: src/components/atoms/AvatarMenu/AvatarMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  readonly user?: User
+  readonly onSignOut?: () => any
+}
+
+interface AvatarMenuState {
+  anchorEl: object
+}
+
+export class AvatarMenu extends React.Component<AvatarMenuProps, AvatarMenuState> {
+  state: { anchorEl: any } = {
```
Comment (lautarodragan) [inline src/components/atoms/AvatarMenu/AvatarMenu.tsx:19]: Shouldn't this be `state: AvatarMenuState`? 

Or maybe not type it at all — don't React typings resolve this by the `React.Component<AvatarMenuProps, AvatarMenuState>`?

File: src/components/molecules/Navbar/Navbar.tsx [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  readonly displaySearch?: boolean
+  readonly searchShadow?: boolean
+  readonly displayNavButtons?: boolean
+  readonly user?: User
+}
+
+const getSearchQuery = () => {
+  const currentLocation = browserHistory.getCurrentLocation()
+
+  if (currentLocation.pathname === '/works') return (currentLocation.query as any).query
```
Comment (lautarodragan) [inline src/components/molecules/Navbar/Navbar.tsx:36]: This doesn't feel right. Components shouldn't access browserHistory or the query params. Can this be provided via props?

@kennylavender @wzalazar for further input

File: src/sagas/SignOut.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  return function*(): IterableIterator<ForkEffect> {
+    yield takeLatest(Actions.SignOut.SIGN_OUT, SignOut)
+  }
+}
+
+export function SignOut(action: any): void {
+  try {
+    if (!(action.payload && action.payload.redirectLogin === false)) browserHistory.push('/')
+  } catch (e) {
+    // Todo: Error message in UI
```
Comment (lautarodragan) [inline src/sagas/SignOut.ts:15]: Should we mention this in the PR description, create an issue and link to it from the PR?

File: src/sagas/SignOut.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+import { browserHistory } from 'react-router'
+import { takeLatest, ForkEffect } from 'redux-saga/effects'
+
+export function SignOutSaga(): () => IterableIterator<ForkEffect> {
+  return function*(): IterableIterator<ForkEffect> {
+    yield takeLatest(Actions.SignOut.SIGN_OUT, SignOut)
+  }
+}
+
+export function SignOut(action: any): void {
```
Comment (lautarodragan) [inline src/sagas/SignOut.ts:11]: can the action be typed?

File: src/actions/index.ts [hunk 9]
```diff
@@ -6,8 +6,12 @@ export namespace Actions {
   export const blockInfoReceived = 'block info received'
 
   export namespace Search {
-    export const Change = 'search change'
-    export const Offset = 'search offset'
+    export const SEARCH_CHANGE = 'SEARCH_CHANGE'
```
Comment (kennylavender) [inline src/actions/index.ts:9]: Do these search constants need to be changed in this PR or can they be adjusted in another PR, I feel like this change is out of scope.

**Comment by kennylavender** [comment]
@jwicks31 I feel like this PR is tackling too many issues at once, I am finding it difficult to keep in mind what your changing/creating and possible issues that could creep in. Can we split it up further. 

For example, something like this, doesn't need to be exactly this.
- install @material-ui/core in its own PR.
- create the sign out reducer, action creator and saga in its own PR.
- create the new storybook component in its own PR
- connect all the things together in one PR after they have been created.

File: src/components/molecules/Navbar/Navbar.tsx (hunk 6 above)
Comment (wzalazar) [reply src/components/molecules/Navbar/Navbar.tsx:36]: yes, I think the same. Another problem is that the component knows about the specific name of our business like /works. I think we have to separate the responsibilities.

Comment (jwicks31) [reply src/components/molecules/Navbar/Navbar.tsx:36]: OK. I kept this function from when I broke apart the Navbar component into a container and component. It was a private method on the original component.

**Comment by jwicks31** [comment]
@kennylavender I'm fine with closing this PR and introducing the changes in smaller PRs.
